### ISSUE TITLE
Saving retained application messages

### ISSUE THREAD

**Issue by elektromntr** [chkr1011/MQTTnet#1206]
### Describe your question
Hi, I've seen in documentation some implementation of Storage for "Saving retained application messages":
https://github.com/chkr1011/MQTTnet/wiki/Server#saving-retained-application-messages

When those methods _SaveRetainedMessagesAsync_ and _LoadRetainedMessagesAsync_ are called?
I have subscribers with persistent sessions and publish messages mostly with QoS > 0

### Which project is your question related to?
- Server

**Comment by BrightSoul** [comment]
Disclaimer: not a Mqttnet dev.

From my experience with the product, the `LoadRetainedMessagesAsync` is [called once by the default MqttRetainedMessageManager](https://github.com/chkr1011/MQTTnet/blob/8f1d4e3c22f3226570eec681bccef3a88c4ebd16/Source/MQTTnet/Server/MqttRetainedMessagesManager.cs#L30) at server startup. In there, you'll have to retrieve ALL retained messages from the database and they will be kept in memory. That's right, if you have 100k retained messages, they will ALL sit in memory. And since each message can be quite large, you could end up wasting a great amount of RAM.

The `SaveRetainedMessagesAsync` is called each time a new persisted message is sent to the server. The method [gets passed the whole list of retained messages](https://github.com/chkr1011/MQTTnet/blob/8f1d4e3c22f3226570eec681bccef3a88c4ebd16/Source/MQTTnet/Server/MqttRetainedMessagesManager.cs#L99) (the aforementioned 100k messages), not just the one that was sent.

After struggling a bit with this, I ended up writing my own implementation of the `IMqttRetainedMessagesManager`. In there, you'll have a chance to load/save only the relevant retained messages.
- In the `GetSubscribedMessagesAsync` method, load and return only the persisted messages that match the provided topic filters;
- In the `HandleMessageAsync` method, persist the provided message in the database;
- Just return `Task.CompletedTask` in all other methods.

**Comment by elektromntr** [comment]
Hi @BrightSoul 
Big big thanks for your reply. It's very helpful.
I have 2 more questions, related to your response:

1. When you say _persisted message_, you mean every message that has QoS not equal to 0?
2. When persisted message is removed from that aforementioned 100k list? When is acknowledged as delivered?

**Comment by BrightSoul** [comment]
Hey @elektromntr 

> When you say persisted message, you mean every message that has QoS not equal to 0?

No no, by _persisted_ I mean a message that's been saved to the database. I think you'll want to save retained messages to the database (i.e. persist them) so that when the application is restarted you can reload them.

> When p̵e̵r̵s̵i̵s̵t̵e̵d̵ retained message is removed from that "aforementioned 100k list"?

No, it's replaced only when another retained message is published to the server for the same topic. As you can [read here](https://www.hivemq.com/blog/mqtt-essentials-part-8-retained-messages/) (emphasis mine):

> A retained message is a normal MQTT message with the retained flag set to true. **The broker stores the last retained message and the corresponding QoS for that topic**.

Mqttnet implements this in its default `MqttRetainedMessageManager` with a `Dictionary<string, MqttApplicationMessage>` whose key is the topic name.
[https://github.com/chkr1011/MQTTnet/blob/8f1d4e3c22f3226570eec681bccef3a88c4ebd16/Source/MQTTnet/Server/MqttRetainedMessagesManager.cs#L82](https://github.com/chkr1011/MQTTnet/blob/8f1d4e3c22f3226570eec681bccef3a88c4ebd16/Source/MQTTnet/Server/MqttRetainedMessagesManager.cs#L82)

And also, it seems the default `MqttRetainedMessageManager` lets you "expire" a previously published retained message by publishing another one with an empty or null payload.
[https://github.com/chkr1011/MQTTnet/blob/8f1d4e3c22f3226570eec681bccef3a88c4ebd16/Source/MQTTnet/Server/MqttRetainedMessagesManager.cs#L75]

...but if you provide your own implementation of a `IMqttRetainedMessagesManager` you'll be able to customize how retained messages are handled and persisted. For instance, this allowed me to persist retained messages in batches, in order to reduce calls to the database.

**Comment by elektromntr** [comment]
Almost everything clear :)

**Correct me if I'm wrong:**
Retained message can is only ONE for one topic. If for example client disconnects disgracefully, and in the meantime will miss 300 retained messages, it will get only ONE last retained message from broker after reconnect?

**Comment by BrightSoul** [comment]
> Retained message is only ONE for one topic. If for example client disconnects disgracefully, and in the meantime will miss 300 retained messages, it will get only ONE last retained message from broker after reconnect?

Exactly but this is not a problem.
Retained messages should carry the full state of the system, not increments/deltas. It's true that, when the client reconnects, it will receive just that last retained message but it's all it needs to accurately reconstruct the UI. That message will contain all the data needed: it's like a snapshot of the current state. All those previous 300 messages are now obsolete and they're not needed anymore.

A few months ago, when I first realized this I thought... well, it's going to be inefficient if I always have to send the full state of the system to my clients. But now I must say it's not really an issue: in my case the full state weighs just a few KBs and I added gzip compression to make it more efficient (message payloads can be binary). I don't use increments/delta messages at all, just retained messages carrying the full state.

**Comment by elektromntr** [comment]
Yeah, it may be the solution.... but... 
in my case, I need every message from publisher to be delivered to each client - I mean whole queue. 
So when client crashes, broker is supposed to deliver historical data to it after reconnect.
And also when broker crashes with some messages in its queues, it has to deliver them to subscribers after restart.

**Comment by elektromntr** [comment]
Yep, that's something like that. 
I do use persistent sessions. I did some tests, and yes, the broker sends messages to the reconnected client that are not yet confirmed. But still no solution for broker's crash. In that case, queues will vanish. Do you know any interface that saves queues and reloads them on "after-crash-restart"?

**Comment by BrightSoul** [comment]
Client sessions are handled by [MqttClientSessionManager](https://github.com/chkr1011/MQTTnet/blob/8f1d4e3c22f3226570eec681bccef3a88c4ebd16/Source/MQTTnet/Server/MqttClientSessionsManager.cs#L19) but since it's a `sealed` class you won't be able to override. It's not implementing any interface either. Perhaps this was not meant to be an extensibility point.

**Comment by elektromntr** [comment]
OK. So I understand that there is no solution yet for persistence of the message queues on broker's level.

**Comment by elektromntr** [comment]
That solution would solve the problem: https://github.com/chkr1011/MQTTnet/issues/498#issuecomment-448552957

**Comment by BrightSoul** [comment]
Yeah, that is interesting. I think you should open a new issue specific to this problem, since that's been closed.
